gateway/api/views.py

```diff
@@ -212,8 +212,33 @@ def upload(self, request):
                 return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
 
             title = serializer.validated_data.get("title")
+            provider_name = serializer.validated_data.get("provider", None)
             author = request.user
-            program = serializer.retrieve_one_by_title(title=title, author=author)
+            if provider_name is None:
+                # Check if title contains the provider: <provider>/<title>
+                logger.debug("Provider is None, check if it is in the title.")
+                provider_name, title = serializer.separate_provider_from_title(
+                    title=title
+                )
+
+            if provider_name:
+                user_has_access = serializer.check_provider_access(
+                    provider_name=provider_name, author=author
+                )
+                if not user_has_access:
+                    # For security we just return a 404 not a 401
+                    return Response(
+                        {"message": f"Provider [{provider_name}] was not found."},
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
+                program = serializer.retrieve_provider_function(
+                    title=title, provider_name=provider_name
+                )
+            else:
```

Comment:
is the logic for this else the same as for the `if provider_name is None` block on L217? if so, we should combine them

Reply:
It's not exactly the same. Maybe it's a bit difficult to see it the idea is the next one:

- this `if` from L217 is to get the `provider_name`, it checks if the user sends a provider, if not we check if the provider is in the title
- `if` from L224 is for when you have definitely the `provider_name`

They are not the same but I think I can move first logic to make it less confuse.

Reply:
WDYT @psschwei ? [04a1c23](https://github.com/Qiskit/qiskit-serverless/pull/1331/commits/04a1c23d55ca6ad99b54e43c413c0b20c2ecb61b)

Reply:
yep, that works